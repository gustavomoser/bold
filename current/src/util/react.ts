import { MutableRefObject, Ref } from 'react'

/**
 * Compose multiple react refs.
 *
 * @returns A new ref that passes down the Element to all ref arguments.
 */
export function composeRefs<T>(...refs: Array<Ref<T>>) {
  return (value: T) => refs.forEach(ref => setRef(ref, value))
}

/**
 * Changes the value of a ref.
 * The ref can be a function ref or a object ref.
 *
 * @param ref The ref to have its value changed.
 * @param value The new value.
 */
export function setRef<T>(ref: Ref<T>, value: any) {
  if (typeof ref === 'function') {
    ref(value)
  } else if (ref) {
    const mutableRef = ref as MutableRefObject<T>
    mutableRef.current = value
  }
}

/**
 * Compose multiple event handlers.
 *
 * @returns A new function that triggers all handlers passed as parameters, in order.
 */
export function composeHandlers(...handlers: Array<(...params: any) => void>) {
  return (...args: any[]) => handlers.forEach(handler => handler && handler(...args))
}

/**
 * Invokes the native `value` property setter of an element.
 * From https://github.com/facebook/react/issues/10135#issuecomment-401496776
 *
 * @param element The element to invoke the setter on.
 * @param value The value to be set.
 */
export function setNativeValue(element: HTMLElement, value: any) {
  const prototype = Object.getPrototypeOf(element)
  const { set: prototypeValueSetter = null } = Object.getOwnPropertyDescriptor(prototype, 'value') || {}
  const { set: valueSetter = null } = Object.getOwnPropertyDescriptor(element, 'value') || {}

  if (prototypeValueSetter && valueSetter !== prototypeValueSetter) {
    prototypeValueSetter.call(element, value)
  } else if (valueSetter) {
    valueSetter.call(element, value)
  } else {
    throw new Error('The given element does not have a value setter')
  }
}
