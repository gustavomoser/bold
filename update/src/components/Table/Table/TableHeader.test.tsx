import { fireEvent, render } from '@testing-library/react'
import React from 'react'

import { Table, TableHead, TableHeader, TableRow } from './index'

describe('TableHeader', () => {
  it('should call onSortChange when clicked', () => {
    const sortChange = jest.fn()
    const { container } = render(
      <Table>
        <TableHead>
          <TableRow>
            <TableHeader sortable={true} sortDirection='ASC' onSortChange={sortChange}>
              Header #1
            </TableHeader>
          </TableRow>
        </TableHead>
      </Table>
    )
    expect(sortChange).not.toHaveBeenCalled()
    fireEvent.click(container.querySelector('th span'))
    expect(sortChange).toHaveBeenCalledWith('DESC', false)
  })
  it('should accept the style prop', () => {
    const { container } = render(
      <Table>
        <TableHead>
          <TableRow>
            <TableHeader style={{ color: 'red' }}>Test</TableHeader>
          </TableRow>
        </TableHead>
      </Table>
    )
    expect(container).toMatchSnapshot()
  })
})
