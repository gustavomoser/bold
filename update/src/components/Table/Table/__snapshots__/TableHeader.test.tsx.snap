// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableHeader should accept the style prop 1`] = `
.emotion-0 {
  width: 100%;
  border-collapse: collapse;
  border: 1px solid #D3D4DD;
}

.emotion-3 {
  text-align: left;
  padding: 0.5rem 1rem;
  border-bottom: 1px solid #D3D4DD;
  color: red;
}

<div>
  <table
    class="emotion-0"
  >
    <thead
      class="emotion-1"
    >
      <tr
        class="emotion-1"
      >
        <th
          class="emotion-3"
        >
          Test
        </th>
      </tr>
    </thead>
  </table>
</div>
`;
